### DISCUSSION TITLE
Publisher animation is not working

### DISCUSSION THREAD

**Discussion by heoblitz** [pointfreeco/swift-composable-architecture#2530] [General]
### Description

In certain cases, the publisher animation is not working, but using run animation solves the problem.

Looking at the logs, it seems that transactions are not being updated correctly when using publisher animations.

If you modify your current animation structure to fire "withTransaction" when you assign a value to a store state, publisher animations will work fine.

```swift
// Store.Swift
//
public func send(
 //...
) -> Task<Void, Never>? {
  // ...
  defer {
    withTransaction(transaction) {
      self.state.value = currentState
    }
  }
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Transaction is updated correctly and the animation works.

### Actual behavior

Transaction doesn't update and the animation doesn't work.

### Steps to reproduce

branch: https://github.com/heoblitz/swift-composable-architecture/tree/issue/animation

file: [swift-composable-architecture.zip](https://github.com/pointfreeco/swift-composable-architecture/files/13101124/swift-composable-architecture.zip)

1. Download the project file and build it with the "AnimationTest Scheme."

2. Proceed with testing while changing the effect return values.

```swift
var body: some Reducer<State, Action> {
  Reduce { state, action in
    switch action {
    case .showDetailButtonTapped:
      // 1. publisher animation is not working
      //
//        return .send(.changeShowingDetailView).animation(.default)
      
      // 2. run animation is working
      //
      return .run { send in
        await send(.changeShowingDetailView)
      }
      .animation(.default)
      
    case .changeShowingDetailView:
      state.isShowingDetailView.toggle()
      return .none
    }
  }
}
```

### 1. publisher animation 

https://github.com/pointfreeco/swift-composable-architecture/assets/25315898/b4b348fd-3890-49d1-8071-c1555fe3acb8

### 2. run animation

https://github.com/pointfreeco/swift-composable-architecture/assets/25315898/257a5b5e-868a-4e4e-a776-588e61630bf5

### The Composable Architecture version information

c472ea4055f9a6ff2e498d35a1ecb83bbda1607e

### Destination operating system

iOS 17.0.3, iOS 15.1

### Xcode version information

Version 15.0 (15A240d)

### Swift Compiler version information

```shell
swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.108 clang-1500.0.40.1)
```

**Comment by raduncicen** [comment]
I am experiencing the same issue

**Comment by mbrandonw** [comment]
Hi @heoblitz, this is happening because effect actions that emit actions immediately (such as is the case with `Effect.send`) are [handled differently](https://github.com/pointfreeco/swift-composable-architecture/blob/9b0f600253f467f61cbd53f60ccc243cc4ff27cd/Sources/ComposableArchitecture/Store.swift#L439-L440) than asynchronous effect actions. The synchronous action is processed immediately in the same call stack as when the original action was sent, and so I don't believe there is a way to process that action with an animation. 

Note that this doesn't have anything to do with `Effect.publisher` vs `Effect.run`. It entirely has to do with synchronous actions from effects. If you update your action like so:

```swift
return .publisher {
  Just(.changeShowingDetailView)
    .delay(for: 0, scheduler: DispatchQueue.main)
}
.animation(.default)
```

…it will also work.

I'm not really sure there is anything the library can do to change this behavior. Our choice to process synchronous actions in the same call stack as when the original action was sent is due to people relying and expecting that kind of behavior. Breaking that expectation will lead to a lot of problems.


I'm inclined to convert this to a discussion since I don't think the library can really do anything to remedy this, and there are simple workarounds. How does that sound?

**Comment by heoblitz** [comment]
@mbrandonw Thanks for the clarification.

It's a minor comment, but it would be nice to have the publisher animation usage in the documentation or comments.

As you say, it would be great to close the issue and turn it into a discussion!
